Scribd shortcode AMP compatibility: allow external links

kienstra opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* In AMP, allow opening external links in the [scribd shortcode](https://wordpress.com/support/scribd/)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Enhances an existing feature

#### Testing instructions:
1. Ensure the [AMP plugin](https://wordpress.org/plugins/amp/) is activte
2. In AMP > Settings, select Transitional mode
3. In Jetpack >Settings > Writing, ensure this is toggled on:

![shortcode-toggle](https://user-images.githubusercontent.com/4063887/79085259-76cdf280-7cfd-11ea-80e5-d0605c47e685.png)

4. Create a new post
5. Add a shortcode block with this, taken from [scribd.php](https://github.com/Automattic/jetpack/blob/feeadafa22309b7cf0d77077c0354d9df1f74eda/modules/shortcodes/scribd.php#L10):

```html
[scribd id=39027960 key=key-3kaiwcjqhtipf25m8tw mode=list]
```

6. Go to the front-end AMP URL:

![amp-view-pre](https://user-images.githubusercontent.com/4063887/79085374-ea6fff80-7cfd-11ea-995b-58b15f453b12.png)

7. In the Scribd shortcode, click an external link
8. Expected: the link opens:

![scribd-external-lnik](https://user-images.githubusercontent.com/4063887/79085540-9a456d00-7cfe-11ea-8590-2fac24078632.gif)

Before this PR, the link didn't open

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Scribd shortcode AMP compatibility improvement for external links

kienstra committed:
Scribd shortcode AMP compatibility: allow iframes to open popups

Add a sandbox attribute to allow opening
external links.

jeherve commented on modules/shortcodes/scribd.php:
```diff
@@ -50,8 +50,9 @@ function scribd_shortcode_handler( $atts ) {
 function scribd_shortcode_markup( $atts ) {
-	$markup = <<<EOD
-<iframe class="scribd_iframe_embed" src="//www.scribd.com/embeds/$atts[id]/content?start_page=1&view_mode=$atts[mode]&access_key=$atts[key]" data-auto-height="true" scrolling="no" id="scribd_$atts[id]" width="100%" height="500" frameborder="0"></iframe>
+	$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ? 'sandbox="allow-popups allow-scripts allow-same-origin"' : '';
+	$markup  = <<<EOD
+<iframe class="scribd_iframe_embed" src="//www.scribd.com/embeds/$atts[id]/content?start_page=1&view_mode=$atts[mode]&access_key=$atts[key]" $sandbox data-auto-height="true" scrolling="no" id="scribd_$atts[id]" width="100%" height="500" frameborder="0"></iframe>
```
Just like for Google Maps, should we take that opportunity to clean things up a bit? We could even make that string translatable, switch everyone to HTTPS, and ensure that link opening in a new window has all the right attributes.

```diff
diff --git a/modules/shortcodes/scribd.php b/modules/shortcodes/scribd.php
index 4cba34662..7bf149fbe 100644
--- a/modules/shortcodes/scribd.php
+++ b/modules/shortcodes/scribd.php
@@ -50,13 +50,24 @@ function scribd_shortcode_handler( $atts ) {
  * @param array $atts Shortcode attributes.
  */
 function scribd_shortcode_markup( $atts ) {
-	$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ? 'sandbox="allow-popups allow-scripts allow-same-origin"' : '';
-	$markup  = <<<EOD
-<iframe class="scribd_iframe_embed" src="//www.scribd.com/embeds/$atts[id]/content?start_page=1&view_mode=$atts[mode]&access_key=$atts[key]" $sandbox data-auto-height="true" scrolling="no" id="scribd_$atts[id]" width="100%" height="500" frameborder="0"></iframe>
-<div style="font-size:10px;text-align:center;width:100%"><a href="https://www.scribd.com/doc/$atts[id]" target="_blank">View this document on Scribd</a></div>
-EOD;
+	$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()
+		? 'sandbox="allow-popups allow-scripts allow-same-origin"'
+		: '';
+	$url     = sprintf(
+		'https://www.scribd.com/embeds/%1$d/content?start_page=1&view_mode=%2$s&access_key=%3$s',
+		absint( $atts['id'] ),
+		esc_attr( $atts['mode'] ),
+		esc_attr( $atts['key'] )
+	);
 
-	return $markup;
+	return sprintf(
+		'<iframe class="scribd_iframe_embed" src="%1$s" %2$s data-auto-height="true" scrolling="no" id="scribd_%3$d" width="100%%" height="500" frameborder="0"></iframe>
+		<div style="font-size:10px;text-align:center;width:100%%"><a href="https://www.scribd.com/doc/%3$d" rel="noopener noreferrer" target="_blank">%4$s</a></div>',
+		esc_url( $url ),
+		$sandbox,
+		absint( $atts['id'] ),
+		esc_html__( 'View this document on Scribd', 'jetpack' )
+	);
 }
 add_shortcode( 'scribd', 'scribd_shortcode_handler' );
 

```

kienstra replied:
Good idea, Jeremy! If it's alright, I'll commit this later tonight.

kienstra replied:
How does https://github.com/Automattic/jetpack/pull/15422/commits/1d847b04459faa4e8be6458a5d3b051fddff84b9 look?

jeherve reviewed (changes requested):
Small suggestion since we're touching this file.

kienstra committed:
Apply Jeremy's suggested improvement to the shortcode render method

Thanks to Alain Schlesser for the logic
to enable testing HTML:
https://github.com/ampproject/amp-wp/blame/1c5d7b23b018142d58caea37c223facf17953905/lib/optimizer/tests/src/MarkupComparison.php#L21-L24

kienstra commented on modules/shortcodes/scribd.php:
```diff
@@ -48,14 +48,35 @@ function scribd_shortcode_handler( $atts ) {
 
-	return $markup;
+	$url = add_query_arg(
+		array(
```
This isn't quite as clean, but this is to avoid URL-encoding `&` in the URL.

For example, with:

```php
	$url     = sprintf(
		'https://www.scribd.com/embeds/%1$d/content?start_page=1&view_mode=%2$s&access_key=%3$s',
		absint( $atts['id'] ),
		esc_attr( $atts['mode'] ),
		esc_attr( $atts['key'] )
	);
```
`$url` can equal `https://www.scribd.com/embeds/39027960/content?start_page=1&view_mode=list&access_key=key-3kaiwcjqhtipf25m8tw`

And when then when it is escaped:
``` 
esc_url( $url ) 
```
...the `&` is URL-encoded:
```html
https://www.scribd.com/embeds/39027960/content?start_page=1&#038;view_mode=list&#038;access_key=key-3kaiwcjqhtipf25m8tw
```

jeherve reviewed (approved):
This tests well for me. 👍

westonruter commented on modules/shortcodes/scribd.php:
```diff
@@ -48,14 +48,35 @@ function scribd_shortcode_handler( $atts ) {
-<iframe class="scribd_iframe_embed" src="//www.scribd.com/embeds/$atts[id]/content?start_page=1&view_mode=$atts[mode]&access_key=$atts[key]" data-auto-height="true" scrolling="no" id="scribd_$atts[id]" width="100%" height="500" frameborder="0"></iframe>
-<div style="font-size:10px;text-align:center;width:100%"><a href="https://www.scribd.com/doc/$atts[id]" target="_blank">View this document on Scribd</a></div>
-EOD;
+	$sandbox = class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request()
+		? 'sandbox="allow-popups allow-scripts allow-same-origin"'
```
It wouldn't hurt to use the same `sandbox` attribute in non-AMP as well. This would eliminate AMP-specific code.
